### PR TITLE
Bug 1841885: Add remapping support for `oc adm catalog mirror`

### PR DESCRIPTION
**Pull request by ecordell**
The second step in performing a "true" airgapped install required
manual manipulation of the output `mapping.txt` to properly
mirror.

The source of `oc adm catalog mirror` is always a catalog image that
has a certain set of image references contained within. When mirroring
these, an ImageContentSourcePolicy and mapping.txt file is generated.

The ICSP always maps the image reference in the catalog to the final
image defined by remapping into the destination registry, so that on
cluster the correct translation can take place.

The mapping.txt file (and the mirroring step performed by oc adm catalog
mirror) always need to map from an existing image to the target image.

When mirroring a catalog to an airgapped cluster, the images need to be
mirrored twice: once to the local filesystem, and once from the local
filesystem.

This commit adds support for a `--remap` flag which can be passed to
the mirror command. This "remaps" the source image found in the catalog
manifests to another location prior to mirroring, so that multi-stage
mirroring is easy.

For example:

```bash
$ oc adm catalog mirror file://rh-ops:1 file://operators --to-manifests=local
$ oc adm catalog mirror file://rh-ops:1 localhost:50000 --remap=./local/mapping.txt
```
This will mirror images from the external registry to the local fs, and
then again from the local fs to the final registry.

The ICSP generated is unmodified, since it will always need to map from
the images defined in the manifests themselves.

### CODE REVIEW

File: pkg/cli/admin/catalog/mirror.go [hunk 1]
```diff
@@ -62,6 +66,13 @@ oc apply -f manifests/imageContentSourcePolicy.yaml
 # Edit the mirroring mappings and mirror with "oc image mirror" manually
 %[1]s --manifests-only quay.io/my/image:latest myregistry.com
 oc image mirror -f manifests/mapping.txt
+
+# Perform an airgapped mirror in two steps:
+# mirror the contents of a catalog to local files and save mapping file in ./to-local
+%[1]s file://my-catalog:latest file://operators --to-manifests=./to-local
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/cli/admin/catalog/mirror.go:72]: nope you're right - I'll update these docs

File: pkg/cli/admin/catalog/mirror.go [hunk 2]
```diff
@@ -62,6 +66,13 @@ oc apply -f manifests/imageContentSourcePolicy.yaml
...
+
+# mirror the contents of a catalog from local files into a target registry using previously calculated mapping file
+%[1]s file://my-catalog:latest my-airgapped-registry:5000 --mapping=./rh-ops-manifests/mapping.txt
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/cli/admin/catalog/mirror.go:75]: You're right, the example is confusing.
  
>   where is the rh-ops-manifests directory coming from?

that's supposed to be the output of the previous command - I can make that clearer

>  is --mapping supposed to be --remap?

yep!

File: pkg/cli/admin/catalog/mirror.go [hunk 3]
```diff
@@ -75,9 +86,11 @@ type MirrorCatalogOptions struct {
 	*IndexImageMirrorerOptions
 	genericclioptions.IOStreams
 
-	DryRun       bool
-	ManifestOnly bool
-	DatabasePath string
+	DryRun         bool
+	ManifestOnly   bool
+	DatabasePath   string
+	RemappingFiles []string
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/cli/admin/catalog/mirror.go:92]: I just did this for parity with with `oc image mirror`, which takes a list of files: https://github.com/openshift/oc/blob/master/pkg/cli/image/mirror/mirror.go#L174

**Comment by smarterclayton** [comment]
Generally mirror commands are expected to understand the context of the mapping.  I.e. `oc adm release mirror` understands the disk format from the image, and you can mirror to disk and then back.  The `--to-` argument should be able to infer a reasonable default from disk or the source.  If mirror.txt is mappings from source to destination (the format mirror would accept), then it's pretty normal to say the user should use `oc image mirror -f mirror.txt` to mirror from disk to final destination.  But if mirror needs to be changed, then usually that would be where you would have that input be provided back to `oc adm catalog mirror`

File: pkg/cli/admin/catalog/mirror.go [hunk 4]
```diff
@@ -48,7 +48,12 @@ var (
 
 		A mapping.txt file is also created that is compatible with "oc image mirror". This may be used to further
 		customize the mirroring configuration, but should not be needed in normal circumstances.
-	`)
+
+		Mirroring into an airgapped environment requires two steps: mirroring from the catalog to local files,
+		and then again from local files into the airgapped registry. See examples ofr the "--remap" flag for more
```
Comment (sallyom) [inline pkg/cli/admin/catalog/mirror.go:53]: ofr/of 
also, I don't see any examples of using --remap flag, you will add those?

**Comment by ecordell** [comment]
closing in favor of #611

**Comment by soltysh** [comment]
> closing in favor of #611

/close

**Pull request closed without merging**